File: src/tools/list-directory.ts
import { z } from 'zod';
import { readdir, stat } from 'fs/promises';
import { join, relative } from 'path';
import type { Tool } from './types';

const ListDirectoryParams = z.object({
  path: z.string().optional().describe('Directory to list. Defaults to current working directory.'),
  recursive: z.boolean().optional().default(false).describe('Whether to list recursively'),
  max_depth: z.number().int().min(1).max(5).optional().default(2).describe('Max recursion depth when recursive=true'),
});

export const listDirectoryTool: Tool = {
  name: 'list_directory',
  description: `List files and directories in a given path. Use this when you need to explore the project structure.

This is the preferred tool for understanding the codebase layout before planning or making changes.
- Returns a clean tree-like structure
- Supports recursive listing with depth control
- Ignores common junk directories (.git, node_modules, dist, etc.) by default`,
  parameters: ListDirectoryParams,
  safety: {
    sideEffect: 'read',
    permission: 'allow',
    reason: 'Lists directory entries without modifying files.',
  },
  async execute(args) {
    const { path = '.', recursive, max_depth } = ListDirectoryParams.parse(args);

    const ignoreDirs = new Set([
      '.git', 'node_modules', 'dist', 'build', '.next', '.turbo', 'coverage',
      '.cache', 'tmp', 'temp', '.DS_Store',
    ]);

    try {
      const root = process.cwd();
      const targetPath = join(root, path);

      const entries = await listDirRecursive(targetPath, root, 0, recursive ? max_depth : 0, ignoreDirs);

      if (entries.length === 0) {
        return `Directory is empty: ${path}`;
      }

      return `Contents of ${path}:\n\n${entries.join('\n')}`;
    } catch (err: any) {
      return `Error listing directory ${path}: ${err.message}`;
    }
  },
};

async function listDirRecursive(
  currentPath: string,
  root: string,
  currentDepth: number,
  maxDepth: number,
  ignoreDirs: Set<string>,
): Promise<string[]> {
  const results: string[] = [];

  try {
    const dirents = await readdir(currentPath, { withFileTypes: true });

    for (const dirent of dirents) {
      const fullPath = join(currentPath, dirent.name);
      const relPath = relative(root, fullPath);

      if (dirent.isDirectory()) {
        if (ignoreDirs.has(dirent.name)) continue;

        const indent = '  '.repeat(currentDepth);
        results.push(`${indent}📁 ${dirent.name}/`);

        if (currentDepth < maxDepth) {
          const children = await listDirRecursive(fullPath, root, currentDepth + 1, maxDepth, ignoreDirs);
          results.push(...children);
        }
      } else {
        const indent = '  '.repeat(currentDepth);
        results.push(`${indent}📄 ${dirent.name}`);
      }
    }
  } catch (e) {
    // ignore permission errors on individual dirs
  }

  return results;
}

File: internal/tools/list_directory.go
package tools

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type listDirectoryTool struct {
	baseTool
	workspaceRoot string
}

func NewListDirectoryTool(workspaceRoot string) Tool {
	return listDirectoryTool{
		baseTool: baseTool{
			name:        "list_directory",
			description: "List files and directories in a workspace path with optional recursion.",
			parameters: Schema{
				Type: "object",
				Properties: map[string]PropertySchema{
					"path":      {Type: "string", Description: "Directory to list. Defaults to workspace root.", Default: "."},
					"recursive": {Type: "boolean", Description: "Whether to list recursively.", Default: false},
					"max_depth": {Type: "integer", Description: "Maximum recursion depth when recursive is true.", Default: 2, Minimum: intPtr(1), Maximum: intPtr(5)},
				},
				AdditionalProperties: false,
			},
			safety: readOnlySafety("Lists directory entries without modifying files."),
		},
		workspaceRoot: normalizeWorkspaceRoot(workspaceRoot),
	}
}

func (tool listDirectoryTool) Run(_ context.Context, args map[string]any) Result {
	// Optional with a "." default: treat an explicit empty path (a common
	// weak-model quirk) the same as the key being absent rather than erroring.
	requestedPath, err := aliasedStringArg(args, []string{"path", "directory", "dir"}, ".", false, true)
	if err != nil {
		return errorResult("Error: Invalid arguments for list_directory: " + err.Error())
	}
	if requestedPath == "" {
		requestedPath = "."
	}
	recursive, err := boolArg(args, "recursive", false)
	if err != nil {
		return errorResult("Error: Invalid arguments for list_directory: " + err.Error())
	}
	maxDepth, err := intArg(args, "max_depth", 2, 1, 5)
	if err != nil {
		return errorResult("Error: Invalid arguments for list_directory: " + err.Error())
	}
	if !recursive {
		maxDepth = 0
	}

	absolutePath, relativePath, err := resolveWorkspacePath(tool.workspaceRoot, requestedPath)
	if err != nil {
		return errorResult("Error listing directory " + requestedPath + ": " + err.Error())
	}

	entries, err := listDirectoryEntries(absolutePath, 0, maxDepth)
	if err != nil {
		return errorResult("Error listing directory " + relativePath + ": " + err.Error())
	}
	if len(entries) == 0 {
		return okResult("Directory is empty: " + relativePath)
	}
	return okResult("Contents of " + relativePath + ":\n\n" + strings.Join(entries, "\n"))
}

func listDirectoryEntries(path string, depth int, maxDepth int) ([]string, error) {
	dirEntries, err := os.ReadDir(path)
	if err != nil {
		return nil, err
	}
	sort.Slice(dirEntries, func(left int, right int) bool {
		if dirEntries[left].IsDir() != dirEntries[right].IsDir() {
			return dirEntries[left].IsDir()
		}
		return dirEntries[left].Name() < dirEntries[right].Name()
	})

	results := make([]string, 0, len(dirEntries))
	for _, entry := range dirEntries {
		if entry.IsDir() && shouldSkipDirectory(entry.Name()) {
			continue
		}

		indent := strings.Repeat("  ", depth)
		if entry.IsDir() {
			results = append(results, indent+entry.Name()+"/")
			if depth < maxDepth {
				children, err := listDirectoryEntries(filepath.Join(path, entry.Name()), depth+1, maxDepth)
				if err == nil {
					results = append(results, children...)
				}
			}
			continue
		}

		results = append(results, fmt.Sprintf("%s%s", indent, entry.Name()))
	}

	return results, nil
}
